pkg/backup/manager.go

```diff
@@ -0,0 +1,171 @@
+	for _, path := range []string{
+		vars.CertRootDir,
+		vars.ManifestsDir,
+		vars.OCIBundleDir,
+		vars.HelmHome,
+		vars.HelmRepositoryConfig,
+	} {
+		logrus.Infof("adding `%s` path to the backup archive", path)
+		bm.Add(NewFilesystemStep(path))
+	}
+	bm.Add(newConfigurationStep(cfgPath))
+}
+
+// Add adds backup step
+func (bm *Manager) Add(step Backuper) {
+	if bm.steps == nil {
+		bm.steps = []Backuper{step}
+		return
+	}
+	bm.steps = append(bm.steps, step)
+}
+
+func (bm Manager) save(backupFileName string, assets []string) error {
+	archiveFile := filepath.Join(bm.tmpDir, backupFileName)
+	logrus.Infof("creating temporary archive file: %v", archiveFile)
+	out, err := os.Create(archiveFile)
+	if err != nil {
+		return fmt.Errorf("error creating archive file: %v", err)
+	}
+	defer out.Close()
+	// Create the archive and write the output to the "out" Writer
+	err = createArchive(out, assets, bm.dataDir)
+	if err != nil {
+		logrus.Fatalf("error creating archive: %v", err)
+	}
+
+	destinationFile := filepath.Join(bm.tmpDir, backupFileName)
+	err = util.FileCopy(archiveFile, destinationFile)
+	if err != nil {
+		return fmt.Errorf("failed to copy archive file from temporary directory: %v", err)
+	}
+	logrus.Infof("archive %s created successfully", destinationFile)
+	return nil
+}
+
+// RunRestore restores cluster
+func (bm *Manager) RunRestore(archivePath string, k0sVars constant.CfgVars) error {
+	if err := util.ExtractArchive(archivePath, bm.tmpDir); err != nil {
+		return fmt.Errorf("failed to unpack backup archive `%s`: %v", archivePath, err)
+	}
+	defer os.RemoveAll(bm.tmpDir)
+	cfg, err := bm.getConfigForRestore(k0sVars)
+	if err != nil {
+		return fmt.Errorf("failed to parse backuped configuration file, check the backup archive: %v", err)
+	}
+	bm.discoverSteps("k0s.yaml", cfg.Spec, k0sVars)
+	for _, step := range bm.steps {
+		if err := step.Restore(bm.tmpDir, bm.dataDir); err != nil {
+			return fmt.Errorf("failed to restore on step `%s`: %v", step.Name(), err)
+		}
+	}
+	return nil
+}
+
+func (bm Manager) getConfigForRestore(k0sVars constant.CfgVars) (*v1beta1.ClusterConfig, error) {
+	configFromBackup := path.Join(bm.tmpDir, "k0s.yaml")
+	_, err := os.Stat(configFromBackup)
+	if os.IsNotExist(err) {
+		return v1beta1.DefaultClusterConfig(k0sVars), nil
+	}
+	logrus.Infof("Using k0s.yaml from: %s", configFromBackup)
+	return v1beta1.FromYamlFile(configFromBackup, k0sVars)
+}
+
+// NewBackupManager builds new manager
+func NewBackupManager() (*Manager, error) {
+	tmpDir, err := ioutil.TempDir("", "k0s-backup")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temporary directory: %v", err)
+	}
+
+	bm := &Manager{
+		tmpDir: tmpDir,
+	}
+
+	return bm, nil
+}
+
+// Backuper defines interface for backup-restore step
+type Backuper interface {
+	Name() string
+	Backup() (StepResult, error)
+	Restore(from, to string) error
+}
+
+// StepResult backup result for the particular step
+type StepResult struct {
```

Comment:
I'd rather make a fields public because now, with public access for Add method, and the interface, it's possible to extend the backup tool by user (like, you can import k0s/pkg/backup and make your own backuper in addition to those we have now)